File: app/src/main/java/com/github/jokar/zhihudaily/widget/LoadLayout.java
package com.github.jokar.zhihudaily.widget;

import android.content.Context;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.github.jokar.zhihudaily.R;

/**
 * Created by JokAr on 2017/7/2.
 */

public class LoadLayout extends LinearLayout {

    ProgressBar progressBar;
    LinearLayout llContainer;
    ImageView imageError;
    TextView tvErrorInfo;
    TextView btnRetry;

    public RetryListener retryListener;

    public LoadLayout(Context context) {
        this(context, null);
    }

    public LoadLayout(Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public LoadLayout(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    void init() {
        LayoutInflater.from(getContext()).inflate(R.layout.layout_load, this, true);
        setOrientation(LinearLayout.VERTICAL);
        setGravity(Gravity.CENTER);

        progressBar = (ProgressBar) findViewById(R.id.progressBar);
        llContainer = (LinearLayout) findViewById(R.id.llContainer);
        imageError = (ImageView) findViewById(R.id.imageError);
        tvErrorInfo = (TextView) findViewById(R.id.tvErrorInfo);
        btnRetry = (TextView) findViewById(R.id.btnRetry);

        btnRetry.setOnClickListener(v -> {
            if (retryListener != null) {
                retryListener.onRetry();
            }
        });

        showLoad();
    }


    public void showLoad() {
        llContainer.setVisibility(View.GONE);
        progressBar.setVisibility(View.VISIBLE);
    }

    public void showError(String error) {
        progressBar.setVisibility(View.GONE);
        llContainer.setVisibility(View.VISIBLE);
        tvErrorInfo.setText(error);
    }

    public interface RetryListener {
        void onRetry();
    }
}

File: app/src/main/java/com/github/jokar/zhihudaily/widget/LoadLayout.kt
package com.github.jokar.zhihudaily.widget

import android.content.Context
import android.util.AttributeSet
import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.ProgressBar
import android.widget.TextView

import com.github.jokar.zhihudaily.R

/**
 * Created by JokAr on 2017/7/1.
 */

class LoadLayout @JvmOverloads constructor(context: Context,
                                           attrs: AttributeSet?,
                                           defStyleAttr: Int = 0) :
        LinearLayout(context, attrs, defStyleAttr) {

    var progressBar: ProgressBar? = null
    var llContainer: LinearLayout? = null
    var imageError: ImageView? = null
    var tvErrorInfo: TextView? = null
    var btnRetry: TextView? = null

    val retryListener: RetryListener? = null

    init {
        init()
    }

    fun init() {
        LayoutInflater.from(context).inflate(R.layout.layout_load, this, true)
        orientation = LinearLayout.VERTICAL
        gravity = Gravity.CENTER


        progressBar = findViewById(R.id.progressBar) as ProgressBar
        llContainer = findViewById(R.id.llContainer) as LinearLayout
        imageError = findViewById(R.id.imageError) as ImageView
        tvErrorInfo = findViewById(R.id.tvErrorInfo) as TextView
        btnRetry = findViewById(R.id.btnRetry) as TextView

        btnRetry?.setOnClickListener {
            retryListener?.onRetry()
        }
    }

    fun showLoad() {
        if (llContainer?.isShown!!) {
            llContainer?.visibility = View.GONE
        }
        progressBar?.visibility = View.VISIBLE
    }

    fun showError(error: String) {
        if (progressBar?.isShown!!) {
            progressBar?.visibility = View.GONE
        }
        llContainer?.visibility = View.VISIBLE
        tvErrorInfo?.text = error
    }

    open interface RetryListener {
        fun onRetry()
    }
}
